cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -1193,23 +1195,52 @@ public void run() {
+     * Sets up the configuration for the task represented by this type.
      *
      * @param conf configuration to modify
+     * @param cConf the cdap configuration
+     * @param runtimeArgs the runtime arguments applicable for this task type
      * @param resources resources information or {@code null} if nothing to set
      */
-    public void setResources(Configuration conf, @Nullable Resources resources) {
+    public void configure(Configuration conf, CConfiguration cConf,
+                          Map<String, String> runtimeArgs, @Nullable Resources resources) {
       if (resources == null) {
         return;
       }
 
-      conf.setInt(memoryConfKey, resources.getMemoryMB());
-      // Also set the Xmx to be smaller than the container memory.
-      conf.set(javaOptsKey, "-Xmx" + (int) (resources.getMemoryMB() * 0.8) + "m");
-
-      if (vcoreConfKey != null) {
+      // Sets the container memory size and vcores if it is not set by in the initialize()
+      if (!isProgrammaticConfig(conf, memoryConfKey)) {
+        conf.setInt(memoryConfKey, resources.getMemoryMB());
+      }
+      if (vcoreConfKey != null && !isProgrammaticConfig(conf, vcoreConfKey)) {
         conf.setInt(vcoreConfKey, resources.getVirtualCores());
       }
+
+      // Setup the jvm opts, which includes options from cConf and the -Xmx setting
+      // Get the default settings from cConf
+      int reservedMemory = cConf.getInt(Configs.Keys.JAVA_RESERVED_MEMORY_MB);
+      double minHeapRatio = cConf.getDouble(Configs.Keys.HEAP_RESERVED_MIN_RATIO);
+
+      // Optionally overrides the settings from the runtime arguments
+      Map<String, String> configs = SystemArguments.getTwillContainerConfigs(runtimeArgs, resources.getMemoryMB());
+      if (configs.containsKey(Configs.Keys.JAVA_RESERVED_MEMORY_MB)) {
+        reservedMemory = Integer.parseInt(configs.get(Configs.Keys.JAVA_RESERVED_MEMORY_MB));
+      }
+      if (configs.containsKey(Configs.Keys.HEAP_RESERVED_MIN_RATIO)) {
+        minHeapRatio = Double.parseDouble(configs.get(Configs.Keys.HEAP_RESERVED_MIN_RATIO));
+      }
+
+      int maxHeapSize = org.apache.twill.internal.utils.Resources.computeMaxHeapSize(resources.getMemoryMB(),
+                                                                                     reservedMemory, minHeapRatio);
+
+      String jvmOpts = cConf.get(Constants.AppFabric.PROGRAM_JVM_OPTS) + " -Xmx" + maxHeapSize + "m";
+      // If the job conf has the jvm opts set programmatically, then prepend it with what being determined above;
+      // otherwise, append. The reason is, we want to have the one set programmatically has higher precedence,
+      // while the one set cluster wide has lower precedence.
+      jvmOpts = isProgrammaticConfig(conf, javaOptsKey)
+        ? jvmOpts + " " + conf.get(javaOptsKey, "")
+        : conf.get(javaOptsKey, "") + " " + jvmOpts;
+      conf.set(javaOptsKey, jvmOpts);
     }
```

Comment:
I think it would be useful to log a debug message including all the settings that were configured in this method. To help users debug the effects of their various settings